BrainPortal/app/controllers/data_providers_controller.rb

```diff
@@ -399,263 +399,328 @@ def browse
-    other_provid = params[:other_data_provider_id] || nil
-    new_dp       = DataProvider.find_accessible_by_user(other_provid,current_user) rescue nil
-    past_tense   = move_or_copy == "MOVE" ? "moved" : "copied"
-    if (move_or_copy == "MOVE" || move_or_copy == "COPY") && !new_dp && !(do_unreg || do_erase)
-      flash[:error] = "Error: you selected to automatically #{move_or_copy} your files but did not specify a destination Data Provider."
-      redirect_to :action => :browse
+    # Is there an automatic copy/move operation to do afterwards?
+    post_action = :copy if params[:auto_do] == "COPY"
+    post_action = :move if params[:auto_do] == "MOVE"
+    target_dp   = DataProvider.find_accessible_by_user(params[:other_data_provider_id], current_user) rescue nil
+    if post_action && ! target_dp
+      flash[:error] = "Missing destination data provider for copy or move."
+      respond_to do |format|
+        format.html { redirect_to :action => :browse }
+        format.xml  { render :xml  => { :error => flash[:error] }, :status => :unprocessable_entity }
+        format.json { render :json => { :error => flash[:error] }, :status => :unprocessable_entity }
+      end
       return
     end
 
-    # Create an association { basename => type } as provided by the form
-    base2type = {}
-    filetypes.select { |typebase| ! typebase.empty? }.each do |typebase|
-      next unless typebase.match(/^(\w+)-(\S+)$/)
-      type = $1
-      base = $2
-      base2type[base] = type
-    end
-
-    # Counters and stats
-    newly_registered_userfiles      = []
-    previously_registered_userfiles = []
-    num_unregistered = 0
-    num_erased       = 0
-    num_skipped      = 0
+    # Provided file types for the file(s) to register. If a file is present
+    # in 'basenames' (to be registered) but *not* in 'filetypes', a default
+    # of 'SingleFile' is used.
+    filetypes = Array(params[:filetypes])
+      .map { |v| [$2, $1] if v.match(/^(\w+)-(\S+)$/) }
+      .compact
+      .to_h
+
+    # Known userfile types, used to validate the values extracted above
+    valid_types = Userfile.descendants
+      .map(&:name)
+
+    # The new file(s)'s default project is the currently active project, if
+    # available.
+    group_id = current_project.try(:id) || current_user.own_group.id
+
+    # Unless one was specified explicitly via :other_group_id
+    group_id = params[:other_group_id].to_i unless
+      params[:other_group_id].blank?
+
+    # Fallback to the user's own project if the one selected above is invalid
+    # (the everyone project, under which no file is ever registered, or a
+    # project the user doesn't have access to).
+    group_id = current_user.own_group.id if (
+      group_id == Group.everyone.id ||
+      ! current_user.available_groups.map(&:id).include?(group_id)
+    )
+
+    # Register the given userfiles in background.
+    userfiles = userfiles_from_basenames(@provider, @as_user, params[:basenames])
```

Comment:
As far as the current code is concerned, `userfiles_from_basenames` returns a hash `{ <basename> => <userfile or nil> }`. As such `userfiles` seems like an appropriate variable name for it.